core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -1071,7 +1079,52 @@ else if (isInferenceCandidate(conjunct)) {
 
                 // Drop predicate after join only if unable to push down to either side
                 if (leftRewrittenConjunct == null && rightRewrittenConjunct == null) {
-                    joinConjuncts.add(allInference.rewrite(conjunct, Sets.union(leftScope, rightScope)));
+                    // Try to push one side of BETWEEN that could not be pushed down entirely.
+                    // We do it only if BETWEEN value is a simple expression, which we assume is cheap to evaluate twice.
+                    if (conjunct instanceof Between between && isSimpleExpression(between.value())) {
```

Comment:
I'm not a big fan of adding this specialized logic for handling Between in what's supposed to be generic pushdown code. It makes the code much harder to maintain and evolve in the long term.

I would suggest looking at splitting a Between into the equivalent comparisons when the value expression is "cheap" in a dedicated IR expression optimizer. Alternatively, look into adding a dedicated Rule for doing this pushdown in the case of a Filter + Join.

Reply:
`between` has nice semantics. Splitting it can have a negative performance impact so I would avoid doing that for the case in this PR as it has limited impact.

A separate rule is a valid alternative. It does lower the cohesiveness of the predicate pushdown logic and makes it harder to understand